File: src/Events/Events/Event.py
from AnalysisG.Particles.Particles import *
from AnalysisG.Templates import EventTemplate


class Event(EventTemplate):
    def __init__(self):
        EventTemplate.__init__(self)
        self.Objects = {
            "Tops": Top(),
            "TopChildren": Children(),
            "TruthJets": TruthJet(),
            "TruthJetPartons": TruthJetParton(),
            "Jets": Jet(),
            "JetPartons": JetParton(),
            "Electrons": Electron(),
            "Muons": Muon(),
        }

        self.Trees = ["nominal"]

        self.weight = "weight_mc"
        self.index = "eventNumber"
        self.mu = "mu"
        self.met = "met_met"
        self.met_phi = "met_phi"

        self.Deprecated = False
        self.CommitHash = "master@7d70c412a65160d897d0253cbb42e5accf2c5bcf"

    def CompileEvent(self):
        self.DetectorObjects = []
        self.Tops = {t.index: t for t in self.Tops.values()}
        self.TopChildren = {
            c.index: c for c in self.TopChildren.values() if isinstance(c.index, int)
        }
        self.TruthJets = {tj.index: tj for tj in self.TruthJets.values()}
        self.TruthJetPartons = {tj.index: tj for tj in self.TruthJetPartons.values()}
        self.Jets = {j.index: j for j in self.Jets.values()}
        self.JetPartons = {j.index: j for j in self.JetPartons.values()}

        for c in self.TopChildren.values():
            self.Tops[c.TopIndex].Children.append(c)
            c.Parent.append(self.Tops[c.TopIndex])
            c.index = c.TopIndex

        for tj in self.TruthJets.values():
            for ti in tj.TopIndex:
                if ti == -1: continue
                tj.Tops += [self.Tops[ti]]
                self.Tops[ti].TruthJets.append(tj)
            tj.index = tj.TopIndex

        for tjp in self.TruthJetPartons.values():
            self.TruthJets[tjp.TruthJetIndex].Parton.append(tjp)
            tjp.Children.append(self.TruthJets[tjp.TruthJetIndex])
            for ci in tjp.TopChildIndex:
                tjp.Parent.append(self.TopChildren[ci])

        for j in self.Jets.values():
            for ti in j.TopIndex:
                if ti == -1: continue
                j.Tops.append(self.Tops[ti])
                self.Tops[ti].Jets.append(j)
            j.index = j.TopIndex

        for jp in self.JetPartons.values():
            self.Jets[jp.JetIndex].Parton.append(jp)
            jp.Children.append(self.Jets[jp.JetIndex])
            for ci in jp.TopChildIndex:
                jp.Parent.append(self.TopChildren[ci])

        maps = {
            i: self.TopChildren[i]
            for i in self.TopChildren
            if self.TopChildren[i].is_lep
        }
        lep = list(self.Electrons.values()) + list(self.Muons.values())
        dist = {maps[i].DeltaR(j): (i, j) for i in maps for j in lep}
        dst = sorted(dist)
        accept = []
        for dr in dst:
            idx, l = dist[dr]
            if l in accept: continue
            maps[idx].Children.append(l)
            l.index = [maps[idx].index]
            l.Parent += maps[idx].Parent
            accept.append(l)

        self.Tops = list(self.Tops.values())
        self.Tops.reverse()
        self.TopChildren = list(self.TopChildren.values())
        self.TopChildren.reverse()
        self.TruthJets = list(self.TruthJets.values())
        self.Jets = list(self.Jets.values())
        self.Electrons = list(self.Electrons.values())
        self.Muons = list(self.Muons.values())
        self.DetectorObjects = self.Jets + self.Electrons + self.Muons

File: src/AnalysisG/events/ssml_mc20/cxx/event.cxx
#include "event.h"

ssml_mc20::ssml_mc20(){
    this -> name = "ssml_mc20"; 
    this -> trees = {"reco"}; 

    this -> add_leaf("eventNumber"   , "eventNumber"); 
    this -> add_leaf("event_category", "event_BkgCategory"); 


    std::string sys = "_NOSYS"; 

    this -> add_leaf("weight"        , "weight_mc"                           + sys); 
    this -> add_leaf("weight_pileup" , "weight_pileup"                       + sys); 
    this -> add_leaf("weight_jvt_sf" , "weight_jvt_effSF"                    + sys); 
    this -> add_leaf("globtrigger_sf", "globalTriggerEffSF"                  + sys); 
    this -> add_leaf("weight_lep_tsf", "weight_leptonSF_tight"               + sys);
    this -> add_leaf("weight_ftag_sf", "weight_ftag_effSF_GN2v01_Continuous" + sys);

    this -> add_leaf("HT_all"   , "HT_all"    + sys);
    this -> add_leaf("ssee"     , "pass_SSee" + sys);   
    this -> add_leaf("ssem"     , "pass_SSem" + sys);   
    this -> add_leaf("ssmm"     , "pass_SSmm" + sys);   
    this -> add_leaf("eem_zveto", "pass_eem_ZVeto" + sys);   
    this -> add_leaf("eee_zveto", "pass_eee_ZVeto" + sys);   
    this -> add_leaf("emm_zveto", "pass_emm_ZVeto" + sys);   
    this -> add_leaf("mmm_zveto", "pass_mmm_ZVeto" + sys);   
    this -> add_leaf("llll_zveto", "pass_llll_ZVeto" + sys);   

    this -> add_leaf("eem", "pass_eem" + sys);   
    this -> add_leaf("eee", "pass_eee" + sys);   
    this -> add_leaf("emm", "pass_emm" + sys);   
    this -> add_leaf("mmm", "pass_mmm" + sys);   

    this -> add_leaf("phi"      , "met_phi"   + sys); 
    this -> add_leaf("met"      , "met_met"   + sys); 
    this -> add_leaf("met_sum"  , "met_sumet" + sys);


    this -> add_leaf("weight_beamsp" , "weight_beamspot"); 

    this -> add_leaf("nElectrons"); 
    this -> add_leaf("nFJets");
    this -> add_leaf("nJets"); 
    this -> add_leaf("nLeptons"); 
    this -> add_leaf("nMuons"); 

    this -> register_particle(&this -> m_zprime); 
    this -> register_particle(&this -> m_tops);
    this -> register_particle(&this -> m_truthjets); 
    this -> register_particle(&this -> m_partons); 
    this -> register_particle(&this -> m_leptons);

    this -> register_particle(&this -> m_electrons); 
    this -> register_particle(&this -> m_muons); 
    this -> register_particle(&this -> m_jets); 
}

ssml_mc20::~ssml_mc20(){}
event_template* ssml_mc20::clone(){return (event_template*)new ssml_mc20();}

void ssml_mc20::build(element_t* el){
    el -> get("eventNumber"   , &this -> eventNumber); 
    el -> get("event_category", &this -> event_category); 

    el -> get("weight"        , &this -> weight_mc); 
    el -> get("weight_beamsp" , &this -> weight_beamspot);
    el -> get("weight_pileup" , &this -> weight_pileup); 
    el -> get("weight_jvt_sf" , &this -> weight_jvt_effSF); 
    el -> get("weight_lep_tsf", &this -> weight_lep_tightSF); 
    el -> get("weight_ftag_sf", &this -> weight_ftag_effSF); 

    el -> get("nElectrons", &this -> n_electrons); 
    el -> get("nFJets"    , &this -> n_fjets); 
    el -> get("nJets"     , &this -> n_jets); 
    el -> get("nLeptons"  , &this -> n_leptons); 
    el -> get("nMuons"    , &this -> n_muons); 

    el -> get("met"       , &this -> met); 
    el -> get("phi"       , &this -> phi); 
    el -> get("met_sum"   , &this -> met_sum); 

    el -> get("globtrigger_sf", &this -> global_trigger_SF); 
    el -> get("HT_all"        , &this -> HT_all);

    char tmp; 
    el -> get("ssee", &tmp); 
    this -> pass_ssee = (int)tmp; 
    el -> get("ssem", &tmp); 
    this -> pass_ssem = (int)tmp; 
    el -> get("ssmm", &tmp); 
    this -> pass_ssmm = (int)tmp; 

    el -> get("eem_zveto", &tmp); 
    this -> pass_eem_zveto = (int)tmp; 
    el -> get("eee_zveto", &tmp); 
    this -> pass_eee_zveto = (int)tmp; 

    el -> get("emm_zveto", &tmp); 
    this -> pass_emm_zveto = (int)tmp; 
    el -> get("mmm_zveto", &tmp); 
    this -> pass_mmm_zveto = (int)tmp; 

    el -> get("eem", &tmp); 
    this -> pass_eem = (int)tmp; 
    el -> get("eee", &tmp); 
    this -> pass_eee = (int)tmp; 

    el -> get("emm", &tmp); 
    this -> pass_emm = (int)tmp; 
    el -> get("mmm", &tmp); 
    this -> pass_mmm = (int)tmp; 

    el -> get("llll_zveto", &tmp); 
    this -> pass_llll_zveto = (int)tmp; 

}

void ssml_mc20::CompileEvent(){
    std::map<int, top*> tops_      = this -> sort_by_index(&this -> m_tops); 
    std::map<int, jet*> jets       = this -> sort_by_index(&this -> m_jets);
    std::map<int, muon*> mux       = this -> sort_by_index(&this -> m_muons); 
    std::map<int, electron*> elx   = this -> sort_by_index(&this -> m_electrons); 
    std::map<int, truthjet*> tjets = this -> sort_by_index(&this -> m_truthjets);
    std::map<int, lepton*>  leps_  = this -> sort_by_index(&this -> m_leptons); 

    this -> vectorize(&this -> m_tops  , &this -> Tops); 
    this -> vectorize(&this -> m_zprime, &this -> Zprime); 
    for (size_t x(0); x < this -> Tops.size(); ++x){
        top* t = (top*)this -> Tops[x]; 
        t -> type = "top";
        if (!t -> from_res){continue;}
        if (!this -> Zprime.size()){continue;}
        this -> Zprime[0] -> type = "zprime";
        this -> Zprime[0] -> register_child(this -> Tops[x]);
        t -> register_parent(this -> Zprime[0]);  
    }
    
    std::map<std::string, parton*>::iterator itp = this -> m_partons.begin(); 
    for (; itp != this -> m_partons.end(); ++itp){
        parton* px = itp -> second; 
        if (px -> jet_index > -1 && jets.count(px -> jet_index)){
            jet* jt = jets[px -> jet_index]; 
            jt -> register_parent(px);
            px -> jets[jt -> hash] = jt; 
        }

        if (px -> truthjet_index > -1 && tjets.count(px -> truthjet_index)){
            truthjet* jx = tjets[px -> truthjet_index]; 
            jx -> register_parent(px);
            px -> truthjets[jx -> hash] = jx; 
            jx -> top_index = px -> top_index; 
        }

        if (px -> muon_index > -1 && mux.count(px -> muon_index)){
            muon* mx = mux[px -> muon_index]; 
            mx -> register_parent(px);
            px -> register_child(mx); 
        }

        if (px -> electron_index > -1 && elx.count(px -> electron_index)){
            electron* ex = elx[px -> electron_index]; 
            ex -> register_parent(px);
            px -> register_child(ex); 
        }
    }

    this -> broken_event += this -> match_object(&tops_, &this -> m_partons); 
    this -> broken_event += this -> match_object(&tops_, &this -> m_electrons); 
    this -> broken_event += this -> match_object(&tops_, &this -> m_muons); 
    this -> broken_event += this -> match_object(&tops_, &this -> m_jets); 
    this -> broken_event += this -> match_object(&tops_, &this -> m_truthjets); 

    this -> vectorize(&this -> m_jets     , &this -> Detector); 
    this -> vectorize(&this -> m_muons    , &this -> Detector); 
    this -> vectorize(&this -> m_electrons, &this -> Detector); 
    this -> vectorize(&leps_, &this -> Leptsn);
    this -> vectorize(&elx  , &this -> Electrons); 
}
